src/theory/quantifiers/sygus/sygus_unif_rl.cpp

```diff
@@ -52,95 +63,161 @@ void SygusUnifRl::notifyEnumeration(Node e, Node v, std::vector<Node>& lemmas)
   lemmas.push_back(exc_lemma);
 }
 
-void SygusUnifRl::addRefLemma(Node lemma)
-{
-  d_prev_rlemmas = d_rlemmas;
-  d_rlemmas = d_tds->getExtRewriter()->extendedRewrite(
-      NodeManager::currentNM()->mkNode(AND, d_rlemmas, lemma));
-  Trace("sygus-unif-rl-lemma")
-      << "SyGuSUnifRl: New collection of ref lemmas is " << d_rlemmas << "\n";
-  d_hasRLemmas = d_rlemmas != d_true;
-}
-
-void SygusUnifRl::collectPoints(Node n)
+Node SygusUnifRl::purifyLemma(Node n,
+                              bool ensureConst,
+                              std::vector<Node>& model_guards,
+                              BoolNodePairMap& cache)
 {
-  std::unordered_set<TNode, TNodeHashFunction> visited;
-  std::unordered_set<TNode, TNodeHashFunction>::iterator it;
-  std::vector<TNode> visit;
-  TNode cur;
-  visit.push_back(n);
-  do
+  Trace("sygus-unif-rl-purify") << "PurifyLemma : " << n << "\n";
+  BoolNodePairMap::const_iterator it = cache.find(BoolNodePair(ensureConst, n));
+  if (it != cache.end())
   {
-    cur = visit.back();
-    visit.pop_back();
-    if (visited.find(cur) != visited.end())
+    Trace("sygus-unif-rl-purify") << "... already visited " << n << "\n";
+    return it->second;
+  }
+  /* Recurse */
+  unsigned size = n.getNumChildren();
+  Kind k = n.getKind();
+  /* Uninterpreted functions (i.e. functions-to-synthesize in general) must be
+     applied over concrete values except top level non-unif function apps */
+  bool fapp = k == APPLY_UF && size > 0 && (ensureConst || usingUnif(n[0]));
+  /* We retrive model value before the next transformation because the resulting
+     node would not have a model value */
+  Node nv = n;
+  /* get model value of non-top level applications of candidates */
+  if (ensureConst && fapp)
+  {
+    nv = d_parent->getModelValue(n);
+    Assert(n != nv);
+    Trace("sygus-unif-rl-purify") << "PurifyLemma : model value for " << n
+                                  << " is " << nv << "\n";
+  }
+  /* Travese to purify */
+  bool childChanged = false;
+  std::vector<Node> children;
+  NodeManager* nm = NodeManager::currentNM();
+  for (unsigned i = 0; i < size; ++i)
+  {
+    if (i == 0 && fapp)
     {
+      children.push_back(n[i]);
       continue;
     }
-    visited.insert(cur);
-    unsigned size = cur.getNumChildren();
-    if (cur.getKind() == APPLY_UF && size > 0)
+    Node child = purifyLemma(n[i], ensureConst || fapp, model_guards, cache);
```

Comment:
I think we technically could make the recursive call be ```!nu_fapp && ( ensureConst || fapp )``` where ```nu_fapp``` is true if we are an evaluation function but it is not a unification one.